Accessibility: Fix overflowing layout on small zoomed screen

samsch opened the pull request:
**What is this feature?**

For the dashboard panel edit view:
When the screen is smaller in height than 500px\* (which does account for zoom!), then
- the layout of the page switches to allowing scrolling over the whole content, ~including the top bar~
- the panel side takes on a minimum height of twice the viewport (giving the panel and query/data section both full screen height
- the panel edit bar no longer scroll its own content, but extends to full height, expanding the page (to then be scrollable)

\* Bikesheddable value! 500px seems about right, but there's a decent argument for 600px as well. *Probably* don't want to go smaller than 500 since at that size usability is already suffering.

**Why do we need this feature?**

UX/Accessibility. Specifically, when the page is zoomed (200% @ 1280x768, equivalent to 640x296) user need to still be able to interact with the content. It does not need to be ideal, only usable.

**Who is this feature for?**

Users using the browser zoom feature

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/107127

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

samsch committed:
fixes the layout in a slightly naive way
does work in both chrome and firefox

samsch committed:
make panel and query sections individually have full viewport height

samsch committed:
allow wrapping in dashboard controls, and align time picker section correctly when wrapped

ashharrison90 commented:
/deploy-to-hg

samsch committed:
use more fixed minimum widths and allow horizontal scroll for overflow

samsch commented:
/deploy-to-hg

samsch committed:
remove collapsing when sizes are fixed, and fix inverted collapse state logic

samsch commented:
/deploy-to-hg

samsch committed:
use new wrapper for reflow layout media query setup

replace the magic numbers with theme breakpoints

apply global styles conditionally and locally

fix left to right splitter collapse state so it's removed in small size

added betterer exception that will be removed in the next commit

samsch committed:
moved component definition outside of non-react class so react hook lint rule recognizes it's not a class component (betterer fixes)

samsch committed:
remove unused import

ashharrison90 reviewed (commented):
think the behaviour in general is pretty good. only minor nit is that by adjusting the header to `absolute` it doesn't fully span the width when scrolling across
<img width="711" height="375" alt="image" src="https://github.com/user-attachments/assets/7ae490f9-d691-4f19-96a9-5948c48beaf9" />
not sure it's worth investing time to fix it 🤷 

now we need to figure out how to make it deliverable 😅 the global css overrides and the splitter style manipulation are the 2 main bits that i'm worried about 🤔 added a couple of comments for discussion

joshhunt commented on public/app/features/dashboard-scene/panel-edit/useScrollReflowLimit.ts:
IMHO:
- hook is fine
- in my view the height is the main constraining factor for this, so if we _had_ to get rid of one, i would do that
- i don't think we should use the small breakpoint value for the height breakpoint. These breakpoints are optimised for width, and so any change to the values certainly would take this into account. I think it's fine to hard code a height value here.

joshhunt reviewed (approved):
LGTM. Just a few minor points, but all together works looks good! Nice job!

eledobleefe reviewed (commented):
My 2 cents on this, I'm missing the bottom border and it makes me feel the content is cut 🤷‍♀️ 

<img width="936" height="652" alt="Captura de pantalla 2025-08-27 a las 16 56 55" src="https://github.com/user-attachments/assets/2e858770-ee7c-4bf9-9f10-9e7a3df3ebd4" />

Apart from that, great work @samsch 🙌

samsch commented on public/app/features/dashboard-scene/panel-edit/PanelEditorRenderer.tsx:
```diff
@@ -80,7 +120,7 @@ function VizAndDataPane({ model }: SceneComponentProps<PanelEditor>) {
   const styles = useStyles2(getStyles);
 
-  const { containerProps, primaryProps, secondaryProps, splitterProps, splitterState, onToggleCollapse } =
+  let { containerProps, primaryProps, secondaryProps, splitterProps, splitterState, onToggleCollapse } =
```
there's a lint rule for this that we aught to have enabled probably. symptom of changing the implementation as i was working on it, thanks for the catch

samsch committed:
nit fix

samsch committed:
move disabling useSnapperSplitter logic into the hook

simplify reflow hook to only use height, and use a fixed height unrelated to shared width breakpoints

samsch commented on public/app/features/dashboard-scene/panel-edit/PanelEditorRenderer.tsx:
```diff
@@ -15,29 +16,68 @@ import { getDashboardSceneFor, getLibraryPanelBehavior } from '../utils/utils';
+
+    primaryProps.className = '';
+    primaryProps.style = {};
+  }
```
Updated to do that!

ashharrison90 reviewed (approved):
lgtm 👍 we should get a review from dashboards-edit as well

samsch committed:
Merge branch 'main' into samsch/adjust-panel-edit-overflow-layout

torkelo commented on public/app/features/dashboard-scene/panel-edit/PanelEditorRenderer.tsx:
```diff
@@ -38,6 +43,25 @@ export function PanelEditorRenderer({ model }: SceneComponentProps<PanelEditor>)
   return (
     <>
+      {isScrollingLayout && (
+        <Global
```
This feels very hacky , to have hook that adds a global css style that overrides global styles.

This should be part of those global styles via plain media queries

torkelo reviewed (commented):
Nice!

Is there a way to make this default css behavior instead (of global styles, components and splitters etc ) vs requiring special dashboard level overrides controlled with hooks

Sergej-Vlasov reviewed (approved):
Tested and looking good!

 We could probably set this absolute header css as a default in `getPageStyles` with media query as it solves the issue when lots of variables completely block the visualisation like in the video below:
 
 

https://github.com/user-attachments/assets/c4983d39-cebf-4c25-9366-b249f609230b


but we definitely need to look into this edge case from dashboards side

torkelo reviewed (changes requested):
Before merging I would like to see if fixes could be more general (ie fixes to fundamental page styles / and components) rather than very specific panel edit overrides

torkelo commented:
@ashharrison90 true it's not very common, but think Explore and a few scene apps use similar page layouts and page splitters

torkelo reviewed (commented):
@ashharrison90 and making the topnav header non sticky feels like a generic behavior all pages should have if we want that responsive behavior

ashharrison90 commented:
@torkelo yeah you're right. we don't think we actually _need_ that global header behaviour for now to satisfy the ticket, so in the interests of getting the other improvements in i'll remove it and we'll figure out if we still want to reintroduce it globally later 👍

ashharrison90 committed:
remove global style overrides

ashharrison90 committed:
Merge branch 'main' into samsch/adjust-panel-edit-overflow-layout

ashharrison90 commented:
/deploy-to-hg

torkelo commented on public/app/features/dashboard-scene/panel-edit/PanelOptionsPane.tsx:
```diff
@@ -178,7 +177,7 @@ function PanelOptionsPaneComponent({ model }: SceneComponentProps<PanelOptionsPa
             )}
           </div>
-          <ScrollContainer>
+          <ScrollContainer minHeight={isScrollingLayout ? 'max-content' : 0}>
```
think we need this PanelDataPaneRendered as well to disable scrolling in side the tab below viz. 

An alternative to this hook => minHeight would be for ScrollContainer to have an option like "disableBelow" or some responsive way to disable scrolling that it can internally control via normal css media query. (vs having multiple useMedia hooks that control react props that intern control css).  

but since it's just here we are using this it's probably not woth to add that yet but something to keep in mind of this scenario keeps coming up in more places.

torkelo reviewed (approved):
looks ok now, not super happy with having to use useMedia hook vs plain css but acceptable for now :) 

but we need to disble scrolling inside the tab below viz as well

ashharrison90 committed:
prevent scrolling in editor

ashharrison90 commented:
> looks ok now, not super happy with having to use useMedia hook vs plain css but acceptable for now :)
> 
> but we need to disble scrolling inside the tab below viz as well

done 👍

dprokop commented:
> If you have enough annotations or variables, then that entire line of controls will fall down instead of breaking in two which leaves blank space until the timepicker

@samsch - this is quite a significant regression in the UI, I think we either the works needs revert or improving so that the initial behavior is restored.

> @mdvictor in your example above, both cases (before/after) would have required 2 llnes of space anyway:

@ashharrison90  - i don' think this is an argument. Case here is not about how many lines the ui takes, but how the controls flow.

torkelo commented:
@ashharrison90 I agree, need to revert or do a quick fix on the controls flow
